[ADD] Project Budget Change Order

patrickrwilson opened the pull request:
Work in progress:
- [x] Need the project_budget module approved and merged as this depends on that module (#578) 
- [x] Change Order Budget Line domain
- [ ] Tests

patrickrwilson committed:
[Add] project_change_order

initial commit

patrickrwilson committed:
[UPD] Added Test

pedrobaeza commented:
Shouldn't this be `project_budget_change_order`?

patrickrwilson commented:
> Shouldn't this be `project_budget_change_order`?

I will make this change, thx

patrickrwilson committed:
[UPD] Name Change

Name change and LINT error fixes.

patrickrwilson committed:
LINT

hieulucky111 commented on project_budget_change_order/data/change_order_stage.xml:
```diff
@@ -0,0 +1,33 @@
+<!-- Copyright 2019 Patrick Wilson <patrickraymondwilson@gmail.com
+    License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+
+<odoo>
```
should be add `noupdate=1`.

patrickrwilson replied:
This has been added.

hieulucky111 commented on project_budget_change_order/models/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from . import project
+from . import change_order
+from . import change_order_stage
+from . import change_order_line
```
order import. (https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#32imports).

patrickrwilson replied:
Fixed

hieulucky111 commented on project_budget_change_order/models/change_order.py:
```diff
@@ -0,0 +1,65 @@
+# Copyright (C) 2019 Pavlov Media
+# License Proprietary. Do not copy, share nor distribute.
+
+from odoo import fields, models, api
```
same above.

patrickrwilson replied:
fixed

hieulucky111 commented on project_budget_change_order/models/change_order.py:
```diff
@@ -0,0 +1,65 @@
+        return self.env.ref(
+            'project_budget_change_order.change_order_stage_draft')
+
+    def action_review(self):
```
should use `api.multi`.

patrickrwilson replied:
Fixed

hieulucky111 commented on project_budget_change_order/models/change_order.py:
```diff
@@ -0,0 +1,65 @@
+        return self.write({'stage_id': self.env.ref(
+            'project_budget_change_order.change_order_stage_review').id})
+
+    def action_draft(self):
```
same above.

patrickrwilson replied:
fixed

hieulucky111 commented on project_budget_change_order/models/change_order.py:
```diff
@@ -0,0 +1,65 @@
+            'project_budget_change_order.change_order_stage_draft').id})
+
+    def action_approve(self):
+        if self.change_order_line_ids:
```
should use `for`.

patrickrwilson replied:
done

hieulucky111 commented on project_budget_change_order/models/change_order.py:
```diff
@@ -0,0 +1,65 @@
+        return self.write({'stage_id': self.env.ref(
+            'project_budget_change_order.change_order_stage_draft').id})
+
+    def action_approve(self):
```
should use `api.multi`.

patrickrwilson replied:
done

hieulucky111 commented on project_budget_change_order/models/change_order.py:
```diff
@@ -0,0 +1,65 @@
+            'project_budget_change_order.change_order_stage_approved').id})
+
+    def action_cancel(self):
+        if self.change_order_line_ids:
```
same above.

patrickrwilson replied:
done and also added for as suggested on the approve method

hieulucky111 commented on project_budget_change_order/models/change_order.py:
```diff
@@ -0,0 +1,65 @@
+
+    @api.multi
+    def action_view_order(self):
+        action = self.env.ref(
```
should add `self.ensure_one()`.

patrickrwilson replied:
done

hieulucky111 commented on project_budget_change_order/models/change_order_stage.py:
```diff
@@ -0,0 +1,22 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models, api
```
order import

patrickrwilson replied:
fixed

hieulucky111 commented on project_budget_change_order/models/project.py:
```diff
@@ -0,0 +1,27 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models, api
```
order import.

patrickrwilson replied:
fixed

hieulucky111 commented on project_budget_change_order/models/project.py:
```diff
@@ -0,0 +1,27 @@
+
+    @api.multi
+    def action_create_change_order(self):
+        action = self.env.ref(
```
should add `self.ensure_one()`.

patrickrwilson replied:
done

hieulucky111 commented on project_budget_change_order/security/project_security.xml:
```diff
@@ -0,0 +1,21 @@
+<!-- Copyright 2019 Patrick Wilson <patrickraymondwilson@gmail.com
+    License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+
+<odoo>
```
should use `noupdate=1`.

patrickrwilson replied:
done

hieulucky111 commented on project_budget_change_order/tests/test_project_change_order.py:
```diff
@@ -0,0 +1,111 @@
+
+from odoo.addons.account_budget_ocs.tests.common import TestAccountBudgetCommon
+
+from datetime import datetime
```
order import

patrickrwilson replied:
This will be updated on the next commit as i'm still working on tests

hieulucky111 reviewed (changes requested):
Please help to check again.

patrickrwilson commented:
@max3903 can you label this as a work in progress as i'm not quite ready for the reviews?

patrickrwilson committed:
[UPD] Fix Reviewer Suggestions

Fix includes reviewer's suggestions

patrickrwilson committed:
[UPD] Missed reviewer suggestions

Fixes a few more reviewer suggestions that were missed.

patrickrwilson commented:
@hieulucky111 thank you for your suggested fixes. code looks much more correct. i think i hit all and the last one for the test will be in the next commit as i've got some other items to do still.

MeetKD commented on project_budget_change_order/models/change_order.py:
```diff
@@ -0,0 +1,72 @@
+    @api.multi
+    def action_cancel(self):
+        for record in self:
+            if record.change_order_line_ids:
```
I guess if condition is  not required here. You can directly iterate on change order line ids

patrickrwilson replied:
Sorry, I'm not quite sure what you mean? Are you saying that the if statement can be removed as it will run correctly even if no budget lines exist?

MeetKD replied:
Yeah.

patrickrwilson replied:
Done, thanks

patrickrwilson committed:
[UPD] Cleanup and more

Added better naming for the account line budget due to having to select those on the change orders. Also some form cleanup and adjustments.

patrickrwilson commented:
I could use some assistance from the community on the last item i'm trying to implement.

On the change order you add change order line items where you select the related budget line to make the adjustment to. I'd like to implement a domain on that field (budget_line_id) to only show budget line items related to the budget that is on the change order. Screenshot below shows this visually.

I've tried adding  domain="[('crossovered_budget_id', '=', change_order_id.budget_id)]" to the budget_line_id field in the change_order_line_ids tree within the change_order.xml however i get the following error: Uncaught Error: AttributeError: object has no attribute 'budget_id'.

I'm not sure what i'm doing wrong. Any suggestions?

![image](https://user-images.githubusercontent.com/36892066/65354600-8c3ecc00-dbb5-11e9-888b-face374180a8.png)

UPDATE: I ended up adding a budget_id field to the change order line and passed the context for that which allowed me to apply the domain. This works well.

pedrobaeza commented:
You can't use dot notation in the right part. There's a little trick in the JS client part that may server: put `domain="[('crossovered_budget_id', '=', parent.budget_id)]"` in the view definition for using the value of the parent model in JS. The tree view must be defined embedded in the view.

patrickrwilson committed:
[UPD] Many Improvements

This includes small changes for readability and adds domain on change order line budget line id.

patrickrwilson commented:
> You can't use dot notation in the right part. There's a little trick in the JS client part that may server: put `domain="[('crossovered_budget_id', '=', parent.budget_id)]"` in the view definition for using the value of the parent model in JS. The tree view must be defined embedded in the view.

Thanks @pedrobaeza , i'll have to look at this method at another time. Your comment did help point me into the direction to get this working by adding the budget_id field to the change order line and using context was able to get the domain to work properly.

patrickrwilson committed:
[UPD] Tests

patrickrwilson committed:
[UPD] ReSync

patrickrwilson committed:
Merge branch '12.0' into project_change_order

patrickrwilson committed:
[FIX] Test py mispelling

patrickrwilson commented:
I'm going to close this as i'm just haven't found the time to work on this and not sure when I can again, i'll open a new one when i have time to complete this module.
